### ISSUE TITLE
screen corruption when using emacs over ssh with conpty off

### ISSUE THREAD

**Issue by cjw296** [jumptrading/FluentTerminal#228]
Unfortunately, even with ConPTY off, I'm seeing screen corruption when using emacs over an ssh connection to as remote machine. I wonder if the trigger is where emacs uses colours?

Steps to reproduce:
- ssh to a remote host
- fire up emacs
- edit a long file in a mode that involves colour highlighting
- scroll down below the end of the first page.

**Comment by cjw296** [comment]
Seems ot affect vim when colours are involved too.

**Comment by peske** [comment]
What kind of corruption happens? I've tested `ssh` -> `emacs some_file.sh`, where the file is bash script with colored syntax, few displays long. I don't see any corruption if I scroll to the end of the file. Also no issues with `ssh` -> `vim some_file.sh`. SSH host is Ubuntu 18.04.

**Comment by mjs** [comment]
@cjw296 is this with or without Mosh? (I'm guessing plain SSH from your description but it's worth confirming)

Can you provide the value of `echo $TERM` on the server in question?

**Comment by cjw296** [comment]
Same as #231, was `xterm-256` but changing it to `xterm` and starting a new session doesn't help.
Like I said above, the key seems to be that terminal colours are involved. If I open a toml file in emacs, everything is fine 'cos there's no syntax highlighting. If I open a perl file, there's syntax highlighting and I have problems where the cursor location jumps around and the screen rendering ends up out of sync.

Could 1903 help with this?

**Comment by MaxRis** [comment]
I've just tried running emacs on RHEL 7.5 with opened perl file. It seems works good.

I've noticed strange cursor position update glitches in emacs when moving through perl file with syntax highlighted and TERM=xterm. If TERM=xterm-256color then these minor cursor glitches don't happen.

**Comment by cjw296** [comment]
I thought we'd been told not to use `xterm-256color` as it caused other, similar problems?

**Comment by MaxRis** [comment]
`xterm-256color` seems has issues with `tmux`.
Just `xterm` worked well for me as well except minor issue with cursor position update time to time.

**Comment by cjw296** [comment]
I'm afraid those cursor position problems are very frequent and quite serious for me...

**Comment by MaxRis** [comment]
Does setting `xterm-256color` resolves these issues for you? (for the record)

**Comment by cjw296** [comment]
Nope. It's not just cursor position rendering, I also get problems with text editing not rendering properly.

**Comment by cjw296** [comment]
Not sure if it's a separate issue, but I also have trouble copying from terminals where there is colour formatting nearby...

**Comment by MaxRis** [comment]
@cjw296 I've tried to repro on perl file of 430 lines and can't see such issues both with TERM set to xterm or xterm-255color.
Might will help if you will share the exact file where you have such issues? or does it happen for any opened file with syntax highlighting?

**Comment by cjw296** [comment]
Anything that involves colour highlighting seems to provoke weirdness. Sadly, it's frustratingly inconsistent.

**Comment by MaxRis** [comment]
@cjw296 could you please post here the `emacs --version` you are using?

I've tried with default from RHEL 7.5:

```
GNU Emacs 24.3.1
Copyright (C) 2013 Free Software Foundation, Inc.
```

and still can't see anything strange. Going to try some newer version of emacs.

**Comment by MaxRis** [comment]
So far tried to repro with

```
GNU Emacs 26.1
Copyright (C) 2018 Free Software Foundation, Inc.
```

And still can't repro the issues.

**Comment by MaxRis** [comment]
In theory, if under some circumstances emacs sends to Fluent too big portions of output (more than  ( NUM_OF_COLUMNS x NUM_OF_ROWS x 4 ) bytes), Fluent might cut the stream accidentally ( as it was happened for `tmux` before the buffer size was not increased ) and really weird visual glitches might happen.

**Comment by mjs** [comment]
> In theory, if under some circumstances emacs sends to Fluent too big portions of output (more than ( NUM_OF_COLUMNS x NUM_OF_ROWS x 4 ) bytes), Fluent might cut the stream accidentally ( as it was happened for `tmux` before the buffer size was not increased ) and really weird visual glitches might happen.

@MaxRis could you try reducing the buffer as an experiment just to see if this could be the issue?

**Comment by MaxRis** [comment]
@mjs @peske 
as discussed in Slack, this source is already replaced with buffered solution. I will try to repro the issue again and also with old TMUX issue condition.

**Comment by MaxRis** [comment]
If to set BufferSize to something minimal like 300 ( Current value of buffer is 204800) screen corruption starts to happen very often on screens with continuous output that exceeds current buffer size. 

![2020-01-19_23-36-42-emacs-screen-corruption](https://user-images.githubusercontent.com/1928902/72689143-13299400-3b17-11ea-8d65-98143aa1cd5e.gif)

Can't repro the issue when buffer equals to 204800.

**Comment by peske** [comment]
No need to reduce the buffer size. The only after effect of huge buffer is slightly bigger memory consumption in SystemTray.

**Comment by mjs** [comment]
I have managed to easily reproduce 2 fairly serious screen corruption issues when using Emacs. The following apply to both:

- FluentTerminal 0.6.1.4
- ssh to a fresh user account on an Ubuntu 18.04 machine  
- ConPTY is off
- using Emacs 25.2.2 without any config (`emacs -q`)
- setting $TERM to `xterm` or `xterm256` doesn't change behaviour 

## Example 1: erratic cursor movement and corruption

See https://youtu.be/-HIbs_Ej3k0 

I opened a new file and started typing some input. After typing the space following `name` the cursor jumps to the end of the line and output continues there. As I type some, more output appears at the correct place but some garbage is also left behind.

## Example 2: corruption after PgDn

See https://youtu.be/WVZwbHTKQ8Y

I opened an existing file with more than a screenful of text and then pressed PgDn. The output is garbled.

**Comment by mjs** [comment]
Here's a dump of what happens when I type 

```
emacs -q test.txt
func hello(name string)
```

The corruption happens after I type the space after `name`.

[20200130103910162_ssh.log](https://github.com/jumptrading/FluentTerminal/files/4133286/20200130103910162_ssh.log)

**Comment by MaxRis** [comment]
I'm getting in log following in case of correctly working emacs:
```
[0m[0K
[52A[2G[?25h[?25l
fu[0K[?25hn[0Kc[0K[6G[?25l
func h[0K[?25he[0Kl[0Kl[0Ko[0K([0Kn[0Ka[0Km[0Ke[0K[17G[?25l
func hello(name s[0K[?25ht[0Kr[0Ki[0Kn[0Kg[0K[?25l
func hello[0;37;46;106m([0mname string)[0K[?25h[?25l
func hello(name string)[0K[12G[?25h[24G[?25l
```

To compare with logs with the issue:
```
[0m[0K
[34A[?25h[2G[?25l
fu[0K[?25hn[0Kc[0K[6G[?25l
func h[0K[?25he[0Kl[0Kl[0Ko[0K([0Kn[0Ka[0Km[0Ke[0K[98G[?25l
func hello(name                                                                                  [0Ks[98G[?25h[?25l
t[0K[?25hr[0Ki[0Kn[0Kg[0K[?25l
[1Afunc hello[0;37;46;106m([0mname string)                                                                          [0Ks[24G[?25h[?25l
func hello(name string)                                                                          [0Ks
[?25h
```

**Comment by MaxRis** [comment]
* The only machine where I can reproduce the issue is 1809 Build 17763.914 XPS15 ( Very odd thing is that I can't repro the issue on VM 1809 Build 17763.316 running on MacBook )
* The same issue is reproducible on the same machine in pure cmd.exe. So, most of all the issue is not related to Fluent
* I have tried to copy OpenSSH executables from the machine with issue to run them on another machine, but this haven't helped to reproduce the issue. It looks like exactly 1809 Build 17763.914 has some serious regression in Window's Conhost.
* Also the issue was NOT reproducible for me on the same XPS15 machine (where the issue is reproducible on 1809 build 17763.914) when running 1903.

**Comment by maxhora** [comment]
As the result of investigation and discussions, such screen corruption might appear in specific versions of Windows 10 and doesn't appear in others (due to internal changes in Windows Conhost and ConPty). Currently, there are no such known issues on latest versions of Windows 10 used by Jump. cc: @cjw296

**Comment by mjs** [comment]
@maxhora shall we close this then?
